Develop Cartesian_Sensor class

anaymehta opened the pull request:
Derives from base class; provides absolute value of net acceleration.

anaymehta committed:
Basic accelerometer code

anaymehta commented on sensors/accelerometer/accelerometer.ino:
```diff
@@ -0,0 +1,25 @@
+  raw_value = sqrt(pow(accelerometer.ax, 2) + pow(accelerometer.ay, 2) + pow(accelerometer.az, 2));
+  m_raw_value = raw_value;
+  sensor_value = sqrt(pow(accelerometer.calcAccel(accelerometer.ax), 2) + pow(accelerometer.calcAccel(accelerometer.ay), 2) + pow(accelerometer.calcAccel(accelerometer.az), 2));
+  m_sensor_value = sensor_value;
```
This value is currently calculated in "g's". I'd argue that m/s^2 is preferable, but I wanted to get a second opinion.

abdullahelqaq replied:
I'm pretty indifferent here as long as it is clear what units we're using. This is the type of information you'll want to have in the readme under `sensors/accelerometer/`

ashvinnagarajan replied:
I like m/s^2 probably. Agreed with Abs on the readme tho

anaymehta commented on sensors/accelerometer/accelerometer.ino:
```diff
@@ -0,0 +1,25 @@
+  // Either use them as raw ADC values, or calculated in g's.
+  raw_value = sqrt(pow(accelerometer.ax, 2) + pow(accelerometer.ay, 2) + pow(accelerometer.az, 2));
+  m_raw_value = raw_value;
+  sensor_value = sqrt(pow(accelerometer.calcAccel(accelerometer.ax), 2) + pow(accelerometer.calcAccel(accelerometer.ay), 2) + pow(accelerometer.calcAccel(accelerometer.az), 2));
```
This line is too long, any ideas on how to shorten it? Maybe a dedicated function to find magnitude from components?

abdullahelqaq replied:
Yeah I'm happy hiding this ugly line behind a function call. Something like `sensor_value = compute_magnitude(accelerometer);` would be neater.

abdullahelqaq commented on sensors/accelerometer.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef ACCELEROMETER_H
+#define ACCELEROMETER_H
```
This is more of a _by the way_ than a request for change. Instead of defining include guards like this, you can also simply use the snappier one liner `#pragma once`. Replaces 3 lines with 1!

abdullahelqaq commented on sensors/accelerometer.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef ACCELEROMETER_H
+#define ACCELEROMETER_H
+
+class Accelerometer: public Sensor
```
I'm sort of confused how you reference `Sensor` without actually including the `Sensor` header or even a forward declaration. Might just be missing something here, but you probably want to `#include` that `Sensor` header.

abdullahelqaq commented on sensors/accelerometer/accelerometer.ino:
```diff
@@ -0,0 +1,25 @@
+  
+  // Now we can use the ax, ay, and az variables as we please.
+  // Either use them as raw ADC values, or calculated in g's.
+  raw_value = sqrt(pow(accelerometer.ax, 2) + pow(accelerometer.ay, 2) + pow(accelerometer.az, 2));
+  m_raw_value = raw_value;
+  sensor_value = sqrt(pow(accelerometer.calcAccel(accelerometer.ax), 2) + pow(accelerometer.calcAccel(accelerometer.ay), 2) + pow(accelerometer.calcAccel(accelerometer.az), 2));
```
Hmm I'm a little concerned that we're losing data here amidst the processing. I'm happy keeping it like this though if we know that we'll never want to use the individual components of the acceleration. 

Generally, this points to a potential shortcoming of our current base sensor implementation, which can store consumable data as a single `double`. What about something like a GPS module that gives us position in 3 different coordinates? Maybe we want a different data structure instead of `double` to hold that in, idk. Thoughts @ashvinnagarajan?

anaymehta replied:
Me and Ash actually discussed it yesterday, and we decided that I would develop a "cartesian_sensor" class derived from "base_sensor" that would include xyz data members. The sensors that currently fall under its scope are the accelerometer, gyroscope, and magnetometer, so it makes sense to develop a single base class for them.

anaymehta replied:
For compute_magnitude() function, will be implemented in the base class as it is a common function to all the sensors.

abdullahelqaq reviewed (changes requested):
A great start! The main change I would like to see is really just the header include. Otherwise, I had more discussion points than I did changes I wanted made. Once we've come to an agreement on the main discussion points I'll be happy to give hit the fat **_Approve_** button 😃

anaymehta committed:
header fixes

abdullahelqaq reviewed (approved):
Looks good to me. I'll approve this, but I would ideally like to see changes made to the README like we mentioned in the review as part of **this PR** so that it lands with the code.

ashvinnagarajan reviewed (commented):
This looks great but we should wait on merging until the cartesian coordinate class comes in too i think

anaymehta committed:
Create readme.md

anaymehta committed:
Cartesian sensor implemented

Implemented Cartesian_Sensor class derived from Sensor, and modified Accelerometer class to correctly derive from Cartesian_Sensor in order to preserve cartesian information.

anaymehta commented:
Requested changes have been addressed!

anaymehta commented:
On a slightly different note, I see that there are two current sensor .ino files in the repository. If it's alright, I'd like to remove one of them to keep things clean.

anaymehta committed:
move .ino files into eponymous folder

Arduino IDE requires the .ino files to be placed in folders

anaymehta committed:
Update README.md

Include sensors directory in readme

anaymehta committed:
Update readme.md

ashvinnagarajan commented on sensors/accelerometer/accelerometer.ino:
```diff
@@ -0,0 +1,24 @@
+
+LSM9DS1 accelerometer;
+
+Accelerometer::Accelerometer(const int &pin): Cartesian_Sensor(&pin) {}
```
Definitely have to initialize our member variables to something. If read_sensor_value goes into the first if statement and then the user attempts to print the sensor value, we would get an undefined error.

ashvinnagarajan reviewed (approved):
Love it looks great!

abdullahelqaq reviewed (approved):
Looks really neat! I agree with the one comment from @ashvinnagarajan, otherwise should be good to merge.

anaymehta committed:
Initialized member variables

Initialized member variables to prevent issues arising from printing.
